---
title: "Microarray Analyses in R"
author: "Jimmy Breen"
date: "01/04/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Microarray's and Public Data

Jimmy Breen  
Bioinformatics Core Facility  
Robinson Research Institute  
University of Adelaide  


**Note:**

This tutorial is _heavily_ influenced by numerous tutorials online:

- http://bioinformatics.knowledgeblog.org/2011/06/20/analysing-microarray-data-in-bioconductor/
- https://www.biostars.org/p/53870/

## Downloading packages

R is a pretty cool programming language to get stuck into gene expression datasets, because there are a tonne of packages that allow you to download data straight off the internet. Most of the really useful packages that facilitate the analysis of biological data are from the [BioConductor](https://bioconductor.org) repository, which is fully open-sourced.

First we'll need to install these Bioconductor packages and load them for our tutorial

```{r download}
# download the BioC installation routines
source("http://bioconductor.org/biocLite.R")

# install the core packages
biocLite()

# install the GEO libraries and some other useful ones
biocLite(c("GEOquery", "ggplot2", "limma", "readr", "magrittr", "oligo", "pd.mogene.2.0.st", "affy", "dplyr", "edgeR", "mogene20sttranscriptcluster.db"))
```

If you've already installed them, then you dont need to do it a second time. You can just skip the last code chunk and go straight to loading the packages

```{r load}
# Load the required libraries
library(GEOquery)
library(ggplot2)
library(dplyr)
library(limma)
library(readr)
library(magrittr)
library(oligo)
library(affy)
library(edgeR)
library(pd.mogene.2.0.st)
library(mogene20sttranscriptcluster.db)
library(org.Mm.eg.db)
library(annotate)
library(biomaRt)
library(pander)
```

## Getting the data

The NCBI GEO database contains all the gene expression 

```{r getStudy}
# ..and then set the directory we want to roll around in
setwd("~/ownCloud/projects/BioinfoHub/microarray_analysis_in_R/GSE70401")

# Lets get johns data using his NCBI ID
getGEOSuppFiles("GSE70401")

# Unpack archive file (the tar file)
untar("GSE70401_RAW.tar", exdir="data")

# List the files in the directory and then unzip each of them in parallel using sapply
cels <- list.files("data/", pattern = "CEL")
sapply(paste("data", cels, sep="/"), gunzip)

```

Ok lets see what we have:

```{r filedir}
cels <- list.files("data/", pattern = "CEL")
```

You should have three files and a directory called "data". The "txt.gz" file is the results published by John and his group. The other two are just left over from us unpacking the file when we downloaded it.

Now we have to give the data some context and specify what each treatment refers to. For this experiment its "INT" (as in "intact") and "SVX" (John to explain these). To do this, we need to add this information into the data object as "group" data.


```{r pheno}
setwd("~/ownCloud/projects/BioinfoHub/microarray_analysis_in_R/GSE70401/data/")
rawData <- read.celfiles(cels)

# Get the samplenames - in this case the filenames of each sample
filename <- sampleNames(rawData)

# Add it was pData
pData(rawData)$filename <- filename

# Those filenames are a bit of a pain in the arse, so lets get rid of the crap at the end of them
sampleNames <- sub("_MoGene-2_0-st_.CEL$", "", filename)
sampleNames(rawData) <- sampleNames

# Now we need to specific what each sample is - intact or svx?
pData(rawData)$group <- c("INT", "INT", "INT", "INT", 
                          "SVX", "SVX", "SVX", "SVX")
pData(rawData)
```

Each sample was basically run individually, so they each have different expression values that have their own biases. To compare them however, we need to normalise them. For this we'll use rma() which is a quantile normalization method that removes background noise in microarrays

```{r normalising}
probesetSummaries <- oligo::rma(rawData, target="probeset")
geneSummaries <- oligo::rma(rawData, target="core")
```


# Quality Control

So we have our data normalised, lets see how the samples relate to each other. We do this by running hierarchial clustering

```{r qc}

#pdf("Hierarchical_clustering.pdf")
m<-t(exprs(rawData))
d <- dist(m)
fit <- hclust(d) 
plot(fit,main="raw data")
m<-t(exprs(probesetSummaries))
d <- dist(m)
fit <- hclust(d) 
plot(fit,main="probe base normalization data")
m<-t(exprs(geneSummaries))
d <- dist(m)
fit <- hclust(d) 
plot(fit,main="gene base normalization data")
#dev.off()
```

Do the samples relate to each other? What clustered with what sample? Do intacts and svx cluster together or separated (like we would expect)?

```{r save_expression}
#make expression matrix file
write.exprs(probesetSummaries,file="normalization_prob_matrix.txt",sep="\t")
write.exprs(geneSummaries,file="normalization_gene_matrix.txt",sep="\t")
```

Lets make a PCA to see how the samples fit together:

```{r pca}
# Make a PCA
color=c('green','green','green','green', 'red','red','red','red')
data.PC = prcomp(t(exprs(geneSummaries)),scale.=TRUE)
plot(data.PC$x[1:8],col=color)

```


# Add Annotation

We need to add gene info to make this all make sense. We're going to put on ENTREZ IDs, which are the gene IDs for NCBI (the main database for genetic data for the entire world).

```{r info}

x <- mogene20sttranscriptclusterENTREZID

# Get the entrez gene IDs that are mapped to an Ensembl ID
mapped_genes <- mappedkeys(x)

# Create dataframes
Annot <- as.data.frame(x[mapped_genes])
df <- data.frame(exprs(geneSummaries)) %>% 
   tibble::rownames_to_column(var="ProbeID")

# Merge data frames together (like a database table join)
all <- dplyr::full_join(Annot, df, by = c("probe_id" = "ProbeID"))


# Write out to a file:
write.table(all, file="normalization_gene_matrix.with_annotation.txt", sep="\t")
```

Now we have a table that has our probe IDs (that dont mean anything other than info )

## Differential gene expression

While this has already been done by John using a different software, I want to show you how to actually quantify the differences between the two treatment groups. To do this, we need to run differential gene expression with a package called "limma".

Limma was developed at the Walter and Elizabeth Hall Institute for Medical Research (WEHI) which has one of the best biostatistical groups in the world. They developed these packages to identify differences between groups within microarrays. They also done a lot of work developing packages to identify differences in RNAseq experiments

```{r deg}

all2 <- read.table("normalization_gene_matrix.with_annotation.txt",
                   sep="\t",
                   head=TRUE,
                   row.names=1)

rownames(all2) <- all2$probe_id

# # Read in our expression data
# x <- read.table("normalization_prob_matrix.txt",sep="\t",head=TRUE,row.names=1) %>%
#    tibble::rownames_to_column(var="probes")


# Bind together the four replicates for each group as intact or svx
# intact <- cbind(x$GSM1726562_JS_I1exp,x$GSM1726563_JS_I2exp, 
#                 x$GSM1726564_JS_I3exp, x$GSM1726565_JS_I4exp)
# 
# svx <- cbind(x$GSM1726566_JS_S1exp, x$GSM1726567_JS_S2exp,
#              x$GSM1726568_JS_S3exp, x$GSM1726569_JS_S4exp)

# Bind together the four replicates for each group as intact or svx
intact <- cbind(all2$GSM1726562_JS_I1exp,all2$GSM1726563_JS_I2exp, 
                all2$GSM1726564_JS_I3exp, all2$GSM1726565_JS_I4exp)

svx <- cbind(all2$GSM1726566_JS_S1exp, all2$GSM1726567_JS_S2exp,
             all2$GSM1726568_JS_S3exp, all2$GSM1726569_JS_S4exp)

# Specific a design matrix model to test
design <- as.matrix(cbind(intact=c(1,1,1,1,0,0,0,0), svx=c(0,0,0,0,1,1,1,1)))
```

The differential expression uses a negative bionomial distribution (shown to be the best reflection of what microarray data looks like) to fit a linear model. Then we identify any significant differences outside that model, and the write our results into a table.


```{r deg2}

x.rma <-cbind(intact,svx)

# Fit the linear model
fit <- lmFit(x.rma,design)

# Define the differences i.e. the treatments
cont.matrix <- makeContrasts(INTvsSVX=svx-intact,levels=design)

# Fit the contrasts between the treatments
fit2 <- contrasts.fit(fit,cont.matrix)
fit2 <- eBayes(fit2)

# Make the gene IDs the rownames
fit2$genes$Symbol=rownames(all2)

# Write the results to a table
write.table(topTable(fit2, adjust="BH",number = 500000),
            file="Int_vs_Svx_gene.txt",
            row.names = TRUE, col.names = TRUE, sep ="\t")

```

Questions:

- What does our results show? How good are results?
- What is the difference between P.Value and adjusted.P.Value?
- Is this, in your opinion, a statistically supported dataset?

## Plots

Data aside, one of the easiests ways to visualise the potentially interesting things within the dataset is to create a volcano plot. These are generally called volcanoes because they look like they're errupting from the ground.

We need to plot the pvalue (usually I take the adjusted.P.Value but we can't in this study) against the log fold change (logFC). Because the P.Values tend to be extremely small, they're often difficult to visualise, so the convention is to negative log transform them.

Here, I'm using the popular package "ggplot2" to plot the values from our results. The "%>%" symbol means a pipe, or a way in which you can pass one command onto the other without making a new datafram or matrix. It's a common way in which old dinosaurs like me use R because it uses the old UNIX commputing style of command-line usage. It also makes R so much easier to use.

Here, I'm removing every gene that is more than a P.Value of 0.01

```{r volano}

topTable(fit2, adjust="BH",number = 500000) %>%
   filter(P.Value < 0.01) %>%
   ggplot(aes(logFC, -log2(P.Value))) +
   geom_point() +
   ylab("Negative log2 (non-corrected) P.Value") +
   xlab("Log Fold Change (logFC)") +
   theme_bw()

```

# Gene Ontology and KEGG Pathways

Lets get a random sample of genes and see if they have enriched Gene Ontology terms and whether they're enriched in any specific pathway. They shouldn't really because they're essentially random, but a few mich pop out

```{r goana_kegg}

deg <- topTable(fit2, adjust="BH",number = 500000) 
rownames(deg) <- deg$Symbol

all2$probe_id <- as.character(all2$probe_id)

new <- deg %>% 
   dplyr::full_join(all2, by = c("Symbol" = "probe_id")) %>%
   dplyr::select(Symbol, gene_id, logFC, AveExpr, P.Value, adj.P.Val) %>%
   filter(P.Value < 0.01)

test_genes <- new[!is.na(new$gene_id),]

test_genes$gene_id %>% 
   goana(species = "Mm", trend = TRUE) %>%
   topGO() %>%
   filter(P.DE < 0.05)

test_genes$gene_id %>%
   kegga(species.KEGG="mmu") %>% 
         arrange(P.DE) %>%
         filter(P.DE < 0.05)

```


Now to get those gene names:

```{r gene_names}

new_noNA <- new[!is.na(new$gene_id),]

mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "mmusculus_gene_ensembl",
  host = 'ensembl.org')

entrez2mgi <- getBM(attributes = c("external_gene_name", "entrezgene", "gene_biotype"),
                    mart = mart,
                    filters = "entrezgene",
                    values = new_noNA$gene_id)

final_table <- dplyr::full_join(new_noNA, entrez2mgi, by = c("gene_id" = "entrezgene")) %>%
   dplyr::select(external_gene_name, Symbol, gene_id, gene_biotype, logFC, AveExpr, P.Value, adj.P.Val)

pander(final_table %>% head(n=20))
```

# Heatmap of differetially expressed genes

The probe IDs are needed to go back and then look at the expression values

```{r gene_heatmap}
filtered <- final_table[!is.na(final_table$external_gene_name),] %>%
   filter(adj.P.Val < 0.1)

# Get the IDs we need to subset on
ids <- filtered$Symbol 

#subset our original dataframe with the expression values and remove the probe/geneID columns
subsetTbl <- all2[rownames(all2) %in% ids,]
subsetTbl <- subsetTbl[,-1]
subsetTbl <- subsetTbl[,-1]
subsetTbl_m <- data.matrix(subsetTbl)

# We want our gene names as rownames
new_rownames <- filtered %>%
   arrange(Symbol) %>%
   dplyr::select(external_gene_name)

row.names(subsetTbl_m) <- new_rownames$external_gene_name

# we also want to get rid of those massive sample names
colnames(subsetTbl_m) <- c("INT1", "INT2", "INT3", "INT4", "SVX1", "SVX2", "SVX3", "SVX4")

library(pheatmap)

pheatmap(subsetTbl_m,
         cutree_rows = 7)
```


# Pathview example

Ideally, what we really want to have is a way of displaying what genes are differentially expressed in one pathway and where they are in the network. Lets see if we can use Pathview to do what we need.

NB: This is an example at this stage. You may need to experiment with what parts of the analysis means (especially the -1 -> 1 heatmap colours)

```{r pathview}

# To install pathview you can run the code below
#source("https://bioconductor.org/biocLite.R")
#biocLite("pathview")

# Load da library
library(pathview)

# We're going to look at one of the pathways that was enriched from our previous KEGG pathway analysis
# - T cell signalling pathway (mmu04660)

pathway_id <- "04660"

# Lets get all of the pathway genes
pathgenes <- getGeneKEGGLinks(species.KEGG = "mmu") %>% 
   filter(PathwayID == "path:mmu04660")

# ...and limit our DEGs to just those in the pathway
deg_path <- final_table[final_table$gene_id %in% pathgenes$GeneID,] 

# Lets subset our "all counts" dataframe to our pathway genes
subsetTbl <- all2[rownames(all2) %in% deg_path$Symbol,]

# Remove the first column because thats just the probe ids
subsetTbl <- subsetTbl[,-1]

# Change the name of our rownames to the entrez gene id (because thats what they use to get the info)
rownames(subsetTbl) <- subsetTbl$gene_id

# Remove that gene id column (because its now our rownames)
subsetTbl <- subsetTbl[,-1]

# convert to a data matrix for plotting
subsetTbl_m <- data.matrix(subsetTbl)

# Run pathview!
pv.out <- pathview(gene.data = subsetTbl_m, gene.idtype="entrez",
                   pathway.id = pathway_id, species = "mmu", min.nnodes = 1,
                   kegg.native = T, same.layer=T,
                   low = list(gene = "SteelBlue", cpd = "blue"),
                   mid = list(gene = "LemonChiffon", cpd = "gray"),
                   high = list(gene = "OrangeRed", cpd = "yellow"))



```


